### PR TITLE
Zero infinity xpu support

### PR DESCRIPTION
**Pull request by Liangliang-Ma**
Currently AIO kernel has gap using on XPU device. 
This PR fixes issues that 
1. cpu buffer only detect CUDA tensor in aio threads 
2. make is_pinned() more general in all kinds of accelerator
3. io_getevents() stopped by signal on XPU

Developed and Tested on Megatron Deepspeed.

### CODE REVIEW

File: csrc/aio/py_lib/deepspeed_aio_thread.cpp
Comment (tjruwase) [inline csrc/aio/py_lib/deepspeed_aio_thread.cpp:36]: Can you clarify that xpu requires 2 copies compared to 1 copy for cuda?

File: csrc/aio/py_lib/deepspeed_py_aio_handle.cpp
Comment (tjruwase) [inline csrc/aio/py_lib/deepspeed_py_aio_handle.cpp:176]: Why disable this assert?

**Comment by Liangliang-Ma** [comment]
Some change seems not necessary. I wonder if it's model related. For example, `assert(_num_pending_ops > 0);` should not raise error if thread's queue works fine. But in Megatron-deepspeed it fails. For `io_pgetevents`, I will test it more to reply you. Thanks.

File: csrc/aio/py_lib/deepspeed_py_aio_handle.cpp
Comment (Liangliang-Ma) [reply csrc/aio/py_lib/deepspeed_py_aio_handle.cpp]: I canceled this disabling. It was model related issue.

**Comment by Liangliang-Ma** [comment]
Hi @tjruwase , I did a modification to current PR.  Removed env variable dependence. Let me explain some change that this PR has.

1) use `io_pgetevents` instead of `io_getevents`
It's indeed more general and it sometimes can help to block signals from interrupting the io process. The performance of these two are same. 

2)More copy on XPU
This is because Async IO requires aligned mem to be used as buffer. Currently it needs manually set to allocate an aligned place, otherwise io would fail. This could be optimized in future.

3) changes in `_swap_out_unpinned_tensors`
In `deepspeed/runtime/swap_tensor/partitioned_optimizer_swapper.py`, `def swap_in_optimizer_state` read data into buffer whose sizes are all `swap_info.numel()`. But for `def _swap_out_unpinned_tensors` in `def swap_out_optimizer_state`, write buffer sizes are numel of each tensor. This could be a mismatch in some cases.

**Comment by tjruwase** [comment]
> 2)More copy on XPU This is because Async IO requires aligned mem to be used as buffer. Currently it needs manually set to allocate an aligned place, otherwise io would fail. This could be optimized in future.
> 
Requiring aligned buffers is a deliberate design decision to simplify the library and ensure high performance.  We prefer for the alignment issues to be addressed on the client side where there is more flexibility. For example, clients can easily add padding in a way that minimizes perf impact. However, I think the library needs to be improved to assert on unaligned buffers to make this requirement more explicit.

**Comment by tjruwase** [comment]
> 3\. changes in `_swap_out_unpinned_tensors`
>     In `deepspeed/runtime/swap_tensor/partitioned_optimizer_swapper.py`, `def swap_in_optimizer_state` read data into buffer whose sizes are all `swap_info.numel()`. But for `def _swap_out_unpinned_tensors` in `def swap_out_optimizer_state`, write buffer sizes are numel of each tensor. This could be a mismatch in some cases.

Can you please share an example of this failure case, perhaps in the form of a unit test?

**Comment by Liangliang-Ma** [comment]
> > 2)More copy on XPU This is because Async IO requires aligned mem to be used as buffer. Currently it needs manually set to allocate an aligned place, otherwise io would fail. This could be optimized in future.
> 
> Requiring aligned buffers is a deliberate design decision to simplify the library and ensure high performance. We prefer for the alignment issues to be addressed on the client side where there is more flexibility. For example, clients can easily add padding in a way that minimizes perf impact. However, I think the library needs to be improved to assert on unaligned buffers to make this requirement more explicit.

The buffers that get used in swap in and out are all created and initialized in deepspeed python codes. In client side we cannot affect it by changing model codes. By the way, I see that in [link](https://github.com/microsoft/DeepSpeed/blob/430510bfce5c8e2e8f2108b73824aef4a37fcaf1/csrc/aio/py_test/ds_aio_handle.py#L34) deepspeed test has a flag that can create aligned tensor which also calls `posix_memalign` to initialize buffer. And this flag seems not get used in deepspeed. What about adding it to let client choose how to create their buffer?

**Comment by Liangliang-Ma** [comment]
> > 3. changes in `_swap_out_unpinned_tensors`
> > In `deepspeed/runtime/swap_tensor/partitioned_optimizer_swapper.py`, `def swap_in_optimizer_state` read data into buffer whose sizes are all `swap_info.numel()`. But for `def _swap_out_unpinned_tensors` in `def swap_out_optimizer_state`, write buffer sizes are numel of each tensor. This could be a mismatch in some cases.
> 
> Can you please share an example of this failure case, perhaps in the form of a unit test?

In my env of Megatron-deepspeed with Adam on xpu devices, when run to `def _update_param_state_info`, the optimizer state that get added to `swap_info` can be a shape of `[1, 22843136, 22843136]`, which will result in mismatch of swap in&out optimizer states. Because in swap_in the buffer is set to `[22843136, 22843136, 22843136]` and `[256, 22843136, 22843136]` in swap_out(aligned).  There is an assert to check this mismatch in [link](https://github.com/microsoft/DeepSpeed/blob/430510bfce5c8e2e8f2108b73824aef4a37fcaf1/csrc/aio/py_lib/deepspeed_py_aio_handle.cpp#L228).

**Comment by Liangliang-Ma** [comment]
@tjruwase Any comments? Thanks!

**Comment by tjruwase** [comment]
> The buffers that get used in swap in and out are all created and initialized in deepspeed python codes. In client side we cannot affect it by changing model codes.

Actually, the expectation is that the client side send aligned buffers to the library. This is why the tensor swapping utility of deepspeed has many alignment [codes](https://github.com/microsoft/DeepSpeed/blob/180dd39714a79b1d30bdecfe5c1b2ffe37b30152/deepspeed/runtime/swap_tensor/optimizer_utils.py#L374) to guarantee this condition.  Can you share a bit about your client side? Perhaps, deepspeed can expose another wapper to provide this alignment. Would that help?

**Comment by tjruwase** [comment]
> By the way, I see that in [link](https://github.com/microsoft/DeepSpeed/blob/430510bfce5c8e2e8f2108b73824aef4a37fcaf1/csrc/aio/py_test/ds_aio_handle.py#L34) deepspeed test has a flag that can create aligned tensor which also calls `posix_memalign` to initialize buffer. And this flag seems not get used in deepspeed. What about adding it to let client choose how to create their buffer?

Yes, good observation. Our plan is to use this functionality within deepspeed and also expose to clients. We hope to do that soon.

**Comment by tjruwase** [comment]
> In my env of Megatron-deepspeed with Adam on xpu devices, when run to `def _update_param_state_info`, the optimizer state that get added to `swap_info` can be a shape of `[1, 22843136, 22843136]`, which will result in mismatch of swap in&out optimizer states. Because in swap_in the buffer is set to `[22843136, 22843136, 22843136]` and `[256, 22843136, 22843136]` in swap_out(aligned).

I am very curious about how this mismatch could happen. It concerns me that our swapping logic is making wrong assumptions. Is it possible to share more details of the training run? Or is it possible to create unit test? Thanks!

**Comment by Liangliang-Ma** [comment]
> I am very curious about how this mismatch could happen. It concerns me that our swapping logic is making wrong assumptions. Is it possible to share more details of the training run? Or is it possible to create unit test? Thanks!

I find the root cause of this mismatch now. The reason is that Deepspeed assumes that optimizer state tensors have same numel and so swap_in&swap_out can use each tensors' numel or param's numel as length. But the truth is different optimizer have different situation. It could be different numel.

For apex.optimizers.FusedAdam, the optimizer states have same numel as param: [apex optimizer states init](https://github.com/NVIDIA/apex/blob/master/apex/optimizers/fused_adam.py#L171)
For torch.optim.Adam, the optimizer states have three tensor and the first one is scalar: [torch adam](https://github.com/pytorch/pytorch/blob/main/torch/optim/adam.py#L98)

I found that my previous change is naive and not enough. So I fallback to use sgd optimizer whose state has same numel as param to bypass this issue. For this part, it can be in another PR or do you want to fix it? @tjruwase

**Comment by tjruwase** [comment]
> I add the alignment in get_accelerator(), which will be easy for different platform to create or adjust their buffer and don't

This solution is fine with me. Thanks!

**Comment by tjruwase** [comment]
> I found that my previous change is naive and not enough. So I fallback to use sgd optimizer whose state has same numel as param to bypass this issue. For this part, it can be in another PR or do you want to fix it? @tjruwase

Thanks for the amazing debugging effort to discover the root issue. I think it is better to fix in another PR. Are you able to help fix this?

File: csrc/aio/py_lib/deepspeed_aio_thread.cpp
Comment (tjruwase) [inline csrc/aio/py_lib/deepspeed_aio_thread.cpp:27]: `|` -> `||`

File: deepspeed/runtime/swap_tensor/async_swapper.py
Comment (tjruwase) [inline deepspeed/runtime/swap_tensor/async_swapper.py:42]: Can you please explain why `is_aligned()` is an alternative for `is_pinned()`? We require `pinned()` buffers in the swapping logic because they are page-locked and so can be used for DMA. Are `aligned()` buffers guaranteed to work with DMA? If so, don't they have to be page-locked as well?

File: accelerator/abstract_accelerator.py
Comment (tjruwase) [inline accelerator/abstract_accelerator.py:225]: After further thought, I think the second API is better because it allows the clients to use other alignments besides the default choice of the accelerator.

File: accelerator/cpu_accelerator.py
Comment (tjruwase) [inline accelerator/cpu_accelerator.py:225]:
```suggestion
    def pin_memory(self, tensor, alignment_bytes=1):
```

**Comment by Liangliang-Ma** [comment]
> After reviewing your alignment changes, I have shared some questions. One idea is that perhaps we could extend the existing `pin_memory()` API to take an alignment flag as well. So that way, clients can request pinned buffers that are also aligned. Also, accelerators would have flexibility in handling such requests. What do you think?
> 
> Below are two possible APIs. The second one seems more general. I am curious if this could help your work. Thanks.

@tjruwase Good idea. In this way we can implement alignment in `pin_memory `and reduce swap code complexity. I choose the first API because I think it's better get `sysconf(_SC_PAGESIZE)` as align bytes rather than let user do the setting. So then `is_align()` is not needed anymore. The code is done.

File: accelerator/cuda_accelerator.py
Comment (tjruwase) [inline accelerator/cuda_accelerator.py:206]:
```suggestion
    def pin_memory(self, tensor, alignment_bytes=1):
```

File: accelerator/mps_accelerator.py
Comment (tjruwase) [inline accelerator/mps_accelerator.py:189]:
```suggestion
    def pin_memory(self, tensor, alignment_bytes=1):
```

File: accelerator/npu_accelerator.py
Comment (tjruwase) [inline accelerator/npu_accelerator.py:191]:
```suggestion
    def pin_memory(self, tensor, alignment_bytes=1):
```

**Comment by Liangliang-Ma** [comment]
> > I found that my previous change is naive and not enough. So I fallback to use sgd optimizer whose state has same numel as param to bypass this issue. For this part, it can be in another PR or do you want to fix it? @tjruwase
> 
> Thanks for the amazing debugging effort to discover the root issue. I think it is better to fix in another PR. Are you able to help fix this?

Yes sure. I can do that.

**Comment by tjruwase** [comment]
> @tjruwase Good idea. In this way we can implement alignment in `pin_memory `and reduce swap code complexity. I choose the first API because I think it's better get `sysconf(_SC_PAGESIZE)` as align bytes rather than let user do the setting. So then `is_align()` is not needed anymore. The code is done.

Sorry :). I think the second API is better so that clients have the option of different alignments, such as super pages.

**Comment by Liangliang-Ma** [comment]
Another thing is that in previous workflow checks, I found that `io_pgetevents` cannot be found in your unittest env when building AIO ops. I am not sure if it's because the testing env is outdated or any other reason. Could you pls help to check with it? @tjruwase 
reference code: [libaio](https://pagure.io/libaio/blob/master/f/src/io_pgetevents.c)

**Comment by tjruwase** [comment]
> It's ok:). I think we can modify it like this to let clients use default page size:
> 
> ```
>     @abc.abstractmethod
>     def pin_memory(self, tensor, align_bytes=1, sc_page_align=False):
> ```

I think it could be confusing to provide two flags to control the alignment bytes. But I also see the value of allowing users to request the default system alignment without having to know the number of bytes. So how about we adjust the second API so that `-1` value for `alignment_bytes` means to use system default such as `sysconf(_SC_PAGESIZE)`? We will document this behavior. What do you think?

**Comment by Liangliang-Ma** [comment]
> I think it could be confusing to provide two flags to control the alignment bytes. But I also see the value of allowing users to request the default system alignment without having to know the number of bytes. So how about we adjust the second API so that `-1` value for `alignment_bytes` means to use system default such as `sysconf(_SC_PAGESIZE)`? We will document this behavior. What do you think?

Yes, I agree. It will be clear more.

**Comment by Liangliang-Ma** [comment]
> > Another thing is that in previous workflow checks, I found that `io_pgetevents` cannot be found in your unittest env when building AIO ops.
> 
> I will pay attention to the new workflow for this issue.

@tjruwase Thanks! btw can you help to trigger workflow once again? Seems it stops when I push new commits.:)

File: csrc/aio/common/deepspeed_aio_common.cpp
Comment (tjruwase) [inline csrc/aio/common/deepspeed_aio_common.cpp:119]:
```suggestion
        aio_ctxt->_io_ctxt, min_completes, max_completes, aio_ctxt->_io_events.data(), nullptr, nullptr);
```

Will this work?

File: deepspeed/runtime/swap_tensor/partitioned_param_swapper.py
Comment (tjruwase) [inline deepspeed/runtime/swap_tensor/partitioned_param_swapper.py:111]:
```suggestion
                                                    align_bytes=0)
```

File: deepspeed/runtime/swap_tensor/utils.py
Comment (tjruwase) [inline deepspeed/runtime/swap_tensor/utils.py:187]:
```suggestion
            get_accelerator().pin_memory(torch.zeros(num_elems, device='cpu', dtype=dtype), align_bytes=0)
```

File: csrc/aio/common/deepspeed_aio_common.cpp
Comment (Liangliang-Ma) [reply csrc/aio/common/deepspeed_aio_common.cpp:119]: Yes, it works. Absolutely.

**Comment by Liangliang-Ma** [comment]
> @Liangliang-Ma, this PR looks good to me after addressing the final suggestions. Thanks so much for this great work.

@tjruwase I committed the change as you suggested. Thanks for your reviewing these days!

**Review by tjruwase — APPROVED** [review]

**Comment by tjruwase** [comment]
> > Another thing is that in previous workflow checks, I found that `io_pgetevents` cannot be found in your unittest env when building AIO ops.
> 
> I will pay attention to the new workflow for this issue.

I think this is failing because the unittest env is ubuntu 18.04 which does not have `io_pgetevents` available. Is it possible for the code to use `io_getevents` or `io_pgetevents` as appropriate? I will also investigate if ubuntu version of the unittest can be updated. Thanks!

**Comment by loadams** [comment]
> > > Another thing is that in previous workflow checks, I found that `io_pgetevents` cannot be found in your unittest env when building AIO ops.
> > 
> > 
> > I will pay attention to the new workflow for this issue.
> 
> I think this is failing because the unittest env is ubuntu 18.04 which does not have `io_pgetevents` available. Is it possible for the code to use `io_getevents` or `io_pgetevents` as appropriate? I will also investigate if ubuntu version of the unittest can be updated. Thanks!

This should be updated to be 20.04 now.  cc: @tjruwase and @Liangliang-Ma 

Is there somewhere we should specify a min libaio version/Ubuntu version required for this?

**Comment by tjruwase** [comment]
@loadams, thanks for updating the CI ubuntu. I think the package version is tied to the OS version, so we need a way to specify minimum Ubuntu version. Unfortunately, I don't know how to do that. I am not sure we have done this previously. 

@mrwyattii, @jeffra any thoughts here? Thanks!

**Comment by tjruwase** [comment]
@Liangliang-Ma, it looks like CI issues are resolved, except for formatting. Can you please address that? Thanks!

**Comment by loadams** [comment]
> @loadams, thanks for updating the CI ubuntu. I think the package version is tied to the OS version, so we need a way to specify minimum Ubuntu version. Unfortunately, I don't know how to do that. I am not sure we have done this previously.
> 
> @mrwyattii, @jeffra any thoughts here? Thanks!

I think we should at a minimum put this in the README in the requirements section?  Also perhaps list the failure mode if one doesn't have this?

**Comment by tjruwase** [comment]
> I think we should at a minimum put this in the README in the requirements section? Also perhaps list the failure mode if one doesn't have this? Thanks! 

Great point @loadams. @Liangliang-Ma, could please replace `io_submit` with `io_pgetevents` in the following compatibility checker function?
https://github.com/microsoft/DeepSpeed/blob/28b9d5c2313addde94a1aa17c09eeb97cb792d7e/op_builder/async_io.py#L76-L82

**Comment by Liangliang-Ma** [comment]
@tjruwase Thanks for solving CI issue! Checker function and formatting modification is committed.

**Comment by Liangliang-Ma** [comment]
@tjruwase seems this PR has been blocked by a cpu inference unit test in merge queue, which is out of this PR's scope. Can you please help to check it? Thanks!

**Comment by loadams** [comment]
> @tjruwase seems this PR has been blocked by a cpu inference unit test in merge queue, which is out of this PR's scope. Can you please help to check it? Thanks!

Hi @Liangliang-Ma - this is a known issue, should be resolved in [this PR](4430) then we can unblock the merge queue.

**Pull request merged**
